File: ch5/clip3/boj16939/main.py
class Cube:
    def __init__(self, area):
        self.area = area

    def check(self):
        for i in range(24):
            base = [0, 4, 8, 12, 16, 20]
            if self.area[i] != self.area[base[i // 4]]:
                return False
        return True

    def r(self):
        temp1 = self.area[1]
        temp2 = self.area[3]
        self.area[1] = self.area[5]
        self.area[3] = self.area[7]
        self.area[5] = self.area[9]
        self.area[7] = self.area[11]
        self.area[9] = self.area[22]
        self.area[11] = self.area[20]
        self.area[22] = temp1
        self.area[20] = temp2

    def u(self):
        temp1 = self.area[4]
        temp2 = self.area[5]
        self.area[4] = self.area[12]
        self.area[5] = self.area[13]
        self.area[12] = self.area[20]
        self.area[13] = self.area[21]
        self.area[20] = self.area[16]
        self.area[21] = self.area[17]
        self.area[16] = temp1
        self.area[17] = temp2

    def f(self):
        temp1 = self.area[2]
        temp2 = self.area[3]
        self.area[2] = self.area[15]
        self.area[3] = self.area[13]
        self.area[15] = self.area[9]
        self.area[13] = self.area[8]
        self.area[9] = self.area[16]
        self.area[8] = self.area[18]
        self.area[16] = temp1
        self.area[18] = temp2

def exit_if_true(b):
    if b:
        print(1)
        exit(0)

# 입력을 받고 Cube 객체를 생성
area = list(map(int, input().split()))
cube = Cube(area)

# 큐브 동작 수행 후 검사
cube.r()
exit_if_true(cube.check())
cube.r()
cube.r()
exit_if_true(cube.check())
cube.r()

cube.u()
exit_if_true(cube.check())
cube.u()
cube.u()
exit_if_true(cube.check())
cube.u()

cube.f()
exit_if_true(cube.check())
cube.f()
cube.f()
exit_if_true(cube.check())

print(0)

File: ch5/clip3/boj16939/Main.java
import java.util.Scanner;

public class Main {
    public static void main(String[] args) {
        Scanner sc = new Scanner(System.in);
        int[] area = new int[25];
        for (int i = 1; i <= 24; i++) {
            area[i] = sc.nextInt();
        }

        Cube cube = new Cube(area);

        cube.r();
        exitIfTrue(cube.check());
        cube.r();
        cube.r();
        exitIfTrue(cube.check());
        cube.r();

        cube.u();
        exitIfTrue(cube.check());
        cube.u();
        cube.u();
        exitIfTrue(cube.check());
        cube.u();

        cube.f();
        exitIfTrue(cube.check());
        cube.f();
        cube.f();
        exitIfTrue(cube.check());

        System.out.println(0);
    }
    static void exitIfTrue(boolean b) {
        if(b) {
            System.out.println(1);
            System.exit(0);
        }
    }
}

class Cube {
    int[] area;
    Cube (int[] area) {
        this.area = area;
    }
    boolean check() {
        for(int i = 1; i <= 24; i++) {
            int[] base = {1, 5, 9, 13, 17, 21};
            if(area[i] != area[base[(i - 1) / 4]]) return false;
        }
        return true;
    }
    void r() {
        int temp1 = area[2];
        int temp2 = area[4];
        area[2] = area[6];
        area[4] = area[8];
        area[6] = area[10];
        area[8] = area[12];
        area[10] = area[23];
        area[12] = area[21];
        area[23] = temp1;
        area[21] = temp2;
    }
    void u() {
        int temp1 = area[5];
        int temp2 = area[6];
        area[5] = area[13];
        area[6] = area[14];
        area[13] = area[21];
        area[14] = area[22];
        area[21] = area[17];
        area[22] = area[18];
        area[17] = temp1;
        area[18] = temp2;
    }
    void f() {
        int temp1 = area[3];
        int temp2 = area[4];
        area[3] = area[16];
        area[4] = area[14];
        area[16] = area[10];
        area[14] = area[9];
        area[10] = area[17];
        area[9] = area[19];
        area[17] = temp1;
        area[19] = temp2;
    }

}